source/shared/cpp/ObjectModel/MarkDownBlockParser.cpp

```diff
@@ -0,0 +1,507 @@
+void EmphasisParser::Flush(std::string& currentToken)
+{
+    if (m_current_state != Text)
+    {
+        CaptureEmphasisToken(currentToken[0], currentToken);
+        m_delimiterCnts = 0;
+    }
+    else
+    {
+        CaptureCurrentCollectedStringAsRegularToken(currentToken);
+    }
+    currentToken.clear();
+}
+
+bool EmphasisParser::IsMarkDownDelimiter(char ch)
+{
+    return ((ch == '*' || ch == '_') && (m_lookBehind != Escape));
+}
+
+void EmphasisParser::CaptureCurrentCollectedStringAsRegularToken(std::string& currentToken) 
+{
+    if (currentToken.empty())
+        return;
+    std::shared_ptr<MarkDownHtmlGenerator> codeGen = 
+        std::make_shared<MarkDownStringHtmlGenerator>(currentToken);
+
+    m_parsedResult.AppendToTokens(codeGen);
+
+    currentToken.clear();
+}
+
+void EmphasisParser::CaptureCurrentCollectedStringAsRegularToken() 
+{
+    CaptureCurrentCollectedStringAsRegularToken(m_current_token);
+}
+
+void EmphasisParser::UpdateCurrentEmphasisRunState(DelimiterType emphasisType)
+{
+    if (m_lookBehind != WhiteSpace)
+    {
+        m_checkLookAhead = (m_lookBehind == Puntuation);
+        m_checkIntraWord = (m_lookBehind == Alphanumeric && emphasisType == Underscore);
+    }
+    ++m_delimiterCnts;
+    m_currentDelimiterType = emphasisType;
+}
+
+bool EmphasisParser::IsRightEmphasisDelimiter(char ch)
+{
+    if (isspace(ch) && 
+       (m_lookBehind != WhiteSpace) && 
+       (m_checkLookAhead || m_checkIntraWord || m_currentDelimiterType == Asterisk))
+    {        
+        return true;
+    }
+
+    if (isalnum(ch) && 
+        m_lookBehind != WhiteSpace && 
+        m_lookBehind != Init)
+    {
+        if (!m_checkLookAhead && !m_checkIntraWord)
+        {
+            return true;
+        }
+
+        if (m_checkLookAhead || m_checkIntraWord)
+        {
+            return false;
+        }
+    }
+
+    if (ispunct(ch) && m_lookBehind != WhiteSpace)
+    {
+        return true;
+    }
+
+    return false;;
+}
+
+bool EmphasisParser::TryCapturingRightEmphasisToken(char ch, std::string &currentToken)
+{
+    if (IsRightEmphasisDelimiter(ch))
+    {
+        std::shared_ptr<MarkDownEmphasisHtmlGenerator> codeGen = nullptr;
+        // right emphasis can be also left emphasis
+        if (IsLeftEmphasisDelimiter(ch))
+        {
+            // since it is both left and right emphasis, create one accordingly
+            codeGen = 
+                std::make_shared<MarkDownLeftAndRightEmphasisHtmlGenerator>(currentToken, m_delimiterCnts, 
+                        m_currentDelimiterType);
+        }
+        else
+        {
+            codeGen = 
+                std::make_shared<MarkDownRightEmphasisHtmlGenerator>(currentToken, m_delimiterCnts, 
+                        m_currentDelimiterType);
+        }
+
+        m_parsedResult.AppendToLookUpTable(codeGen);
+
+        std::shared_ptr<MarkDownHtmlGenerator> htmlToken = std::static_pointer_cast<MarkDownHtmlGenerator>(codeGen);
```

Comment:
>std::shared_ptr<MarkDownHtmlGenerator> htmlToken = std::static_pointer_cast<MarkDownHtmlGenerator>(codeGen); [](start = 8, length = 108)

You don't need this cast.  If you update AppendToTokens to take a const reference you can just do m_parsedResult.AppendToTokens(codeGen);

Reply:
thanks! will update that

---
In reply to: [157626428](https://github.com/Microsoft/AdaptiveCards/pull/1045#discussion_r157626428) [](ancestors = 157626428)